### ISSUE TITLE
Support more encodings ("charset" property)

### ISSUE THREAD

**Issue by kAlvaro** [editorconfig/editorconfig#209]
I suppose there're technical reasons to make `charset` support only a closed set of encodings but current list falls pretty short. I'm specially missing encodings that are common in my area (Western Europe):

* ISO-8859-15 (Latin-9) - *Latin-1's fork with € symbol*
* Windows-1252 - *Default encoding in Windows computers sold here*

Use cases:

* Ensure a legacy project doesn't get messed by a modern "UTF-8 by default" IDE
* Encode helper scripts (*.bat) in native format

Other countries will probably miss other encodings.

If you need a list of encodings and aliases you can probably borrow it from the [iconv project](http://www.gnu.org/software/libiconv/).

**Comment by jedmao** [comment]
It looks like the [iconv-lite](https://github.com/ashtuchkin/iconv-lite/wiki/Supported-Encodings) npm package covers those encodings. I'll make sure to work these into the [ECLint project](https://github.com/jedmao/eclint).

**Comment by xuhdev** [comment]
@jedmao I would like to point out that it's better that all encoding follows the names defined by [IANA](http://www.iana.org/assignments/character-sets/character-sets.xhtml), thus editors have a better chance to convert those names to their internal encoding names, and we would also have a defined behavior across different libraries/editors.

**Comment by jedmao** [comment]
Thanks @xuhdev that will actually be quite helpful.

**Comment by kAlvaro** [comment]
Using IANA preferred names looks sensible but it's worth noting that they'll break backwards compatibility:

- `latin1` should be `ISO-8859-1` (or `ISO_8859-1:1987`? I don't really understand the [table](http://www.iana.org/assignments/character-sets/character-sets.xhtml#table-character-sets-1))
- `utf-8-bom` has no equivalent because, after all, it isn't an encoding per-se

Other current values are fine, assuming case doesn't matter:

- `utf-8`
- `utf-16be`
- `utf-16le`

Of course, they can always be left as documented exceptions.

**Comment by xuhdev** [comment]
I think `latin1` should be `US-ASCII`. Any way, they can be kept as exceptions.

**Comment by jedmao** [comment]
This brings up another question. How should we be determining that a document is of a particular encoding? Sure, the .editorconfig file says it **should** be a certain encoding, but that doesn't mean the document is already in that encoding.

There are a couple of ways to do it:
1. Rely on BOM signatures to detect encoding.
2. Scan the contents of the document to infer encoding (not 100% accurate and potentially slow).

Yes, there are libraries that assist in detecting charsets, but these are the methods they would use.

I'm thinking we can only really rely on the BOM signature, but that leaves unsigned documents as undetectable.

Anyone have some thoughts on this?

**Comment by jedmao** [comment]
Specifically, I'm talking about how [ECLint](https://github.com/jedmao/eclint) should handle the charsets. Should I detect the document charset and convert it to the appropriate one? That's a pretty complicated feat!

I do have it working that way for BOM-signed docs, but I don't really want to go to much deeper. I could really use some advice.

**Comment by hax** [comment]
WHATWG Encoding spec deprecates all legacy encodings include utf-16 and latin1.  Though this spec is mainly for Web browsers, it's has important reference value. 

New projects could stick on utf-8. And it's easy to do one-off encoding conversion for the legacy projects.

So I think it may be not very useful to support other encodings. Especially it's no reliable and good performence way to detect ASCII compatible legacy encodings like ISO-8859-X, GBK, Shift-JIS, etc.

**Comment by randstr** [comment]
> I do have it working that way for BOM-signed docs, but I don't really want to go to much deeper. I could really use some advice.

Encoding detection is full of pitfalls. Either the user gives the encoding or leave that mess to the editor  IMHO.

**Comment by kAlvaro** [comment]
A question from someone who's unaware of the project internals... Is it important for the overall logic to be able to auto-detect every property that's implemented?

I don't think that UTF-8 needs to be particularly enforced since it's the only sensible default for new projects. The main value of the *charset* property lies on legacy codebases, where encoding can typically be a random one and cannot be easily detected. (Plus .bat files are no longer reliable if written in UTF-8.)

**Comment by hax** [comment]
> (Plus .bat files are no longer reliable if written in UTF-8.)

Why? I have no idea about it.

**Comment by kAlvaro** [comment]
@hax I can't point you to proper documentation (sorry) but every time I forget to save a .bat file as ANSI I start getting seemingly random malfunctions, including some commands don't executing at all (so it isn't an urban legend, I've suffered it myself as late as 2017 on Windows 10 x64). I sometimes run `chcp 65001` to be able to read output from UTF-8 PHP console scripts with nothing but minor glitches (such as missing or duplicate characters) but actually saving batch files as UTF-8 tends to be catastrophic.

**Comment by xuhdev** [comment]
@kAlvaro Is [this](http://stackoverflow.com/a/388500/1150462) what you want?

**Comment by kAlvaro** [comment]
@xuhdev  While it's still not official documentation (or documentation), several user comments state the same I do.

**Comment by hax** [comment]
@kAlvaro As the discussion of stackoverflow, you need use `chcp` to change code page to utf8 (65001). Anyway, old BAT is not unicode aware, and should always use simple ASCII chars, and in that case, there is no difference with ANSI and UTF8.

**Comment by raholly** [comment]
@xuhdev So,will we support more encodings?Our code files are all encoding by gbk,then we can only abandon editorconfig now,but that's not what we want.

**Comment by xuhdev** [comment]
@raholly I don't think we have plans in the near future. Encodings use very different names in different systems, which is hard to map in the plugins. In addition, I think you might want to consider converting your source code to unicode -- it's better supported than GBK in developers' tools.

**Comment by AlexeiScherbakov** [comment]
Some very old code C is needed to be in ANSI. For example i have one very old project with old Analog Devices DSP with is needed to be in Windows 1251 codepage (8-bit Cyrillic encoding).

I think that Windows-125x encodings is widely used in old code.

**Comment by binki** [comment]
I looked at https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#charset and noticed that it doesn’t mention `utf-8-bom`. But http://editorconfig.org/#supported-properties does mention is. Why is this inconsistent?

**Comment by binki** [comment]
@xuhdev Thanks. Sorry for the bugspam. I assume I shouldn’t create a new issue since you went and fixed it. It’s OK to delete my comments on this issue if that helps.

**Comment by flowl** [comment]
I am getting errors like `EditorConfig: "windows-1252" is not a valid value for charset for file xy.html` in phpStorm.
Hasn't this been fixed yet?

**Comment by kAlvaro** [comment]
@flowl This ticket remains open and [docs](https://editorconfig.org/#supported-properties) still say *&laquo;charset: set to "latin1", "utf-8", "utf-8-bom", "utf-16be" or "utf-16le" to control the character set&raquo;* so I guess it hasn't been fixed.

This looks plain simple to the untrained eye but apparently there're rooted technical issues that makes this more complex that it looks.

**Comment by flowl** [comment]
Not beeing able to set common encodings makes a big part of .editorconfig obsolete.
We could open a bug at [Jetbrains](https://youtrack.jetbrains.com/issues?q=tag:%20EditorConfig) but it doesn't make sense until .editorconfig "officially" is supposed to support generic encodings. Why can't those simple words not simply be added to the documentation within three years. Is this project even active?

**Comment by xuhdev** [comment]
@kAlvaro is right: While this looks simple, practically it is very hard to work out. All editor plugin developers have to create a mapping from our set encoding names to the names used internally in their respective editors. How do you do it if editors do not commonly provide such a map to "standard" encoding names? For example, I don't think many other editors recognize "windows-1252." And, additionally, how do we know which non-mainstream encodings are widely supported?

**Comment by xuhdev** [comment]
If you are only requesting to support windows-1252, [I think practically you can use latin1](https://msdn.microsoft.com/en-us/library/cc195054.aspx). It should be more or less the same as windows-1252, but under a different name. But, supporting a huge set of encodings, as stated before, isn't practical as editors use different encoding names internally.

**Comment by flowl** [comment]
We are using a lot of characters from windows-1252 which are not represented the same in latin1, like tildes and several different dashes, it's not an option :-/ 
But how do you think we can start to support more encodings when the .editorconfig implementation interface does not require supporting more encodings? I don't understand how you try to solve this. Besides that, .editorconfig is really nice and something that should have been there since the beginning of IDE programing <3

**Comment by ppalaga** [comment]
@xuhdev `latin1` is a common alias for `iso-8859-1`. `iso-8859-1` is different enough from `windows-1252` for not being practically interchangeable with each other. See e.g. http://string-functions.com/encodingtable.aspx?encoding=28591&decoding=1252

**Comment by flowl** [comment]
Just because the internal identifiers for encodings differ from editor to editor is not a reason to not support them. This is not a really problem, put up a mapping, done. It's like putting a few words in an array. Why do you see a problem in this?

**Comment by ppalaga** [comment]
For the individual EditorConfig processors this is a question of having some sort of a support for encodings (and their aliases) in their ecosystem. I guess every modern language has means at its disposal to handle most of the common encodings and their aliases. But still, the sets of supported encoding names will vary among languages.

Clearly, the end users of `.editorconfig` files need to know what the supported encoding names are. But because the sets of supported encoding names are out of EditorConfig organization's control, we cannot say more than "Dear user, go and check the documentation of the EditorConfig processing tools you use".

Having said that, I propose the following:

1. Authors of EditorConfig processing tools will be encouraged to accept every possible encoding name they understand. If an EditorConfig processing tool encounters an encoding name it does not understand, it should output the list (or a link to a list) of supported encoding names.

2. The docs of the `charset` property [1]  and [2] will be rephrased in the following way:

-----8<-------------

`charset` - The name of the character encoding to use for reading from or writing to the selected files.
Character encoding names are case insensitive. The exact set of possible values depends on the software you use for the processing of `.editorconfig` files, so you should refer to its documentation. However, `.editorconfig` processors should support at least the following common encoding names standardized by [IANA](https://www.iana.org/assignments/character-sets/character-sets.xhtml):

utf-8
utf-8-bom
utf-16
utf-16be
utf-16le
us-ascii
iso-8859-1
latin1
iso-8859-2
iso-8859-3
iso-8859-4
iso-8859-5
iso-8859-6
iso-8859-7
iso-8859-8
iso-8859-9
iso-8859-13
iso-8859-14
iso-8859-15
shift_jis
euc-jp
iso-2022-kr
euc-kr
iso-2022-jp
iso-2022-jp-2
gb2312
big5
koi8-r

-----8<-------------

What do you think?

[1] https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#charset 
[2] https://editorconfig.org/#supported-properties

**Comment by ffes** [comment]
In general I don't have any problems with extending the list. If people really need this, why not? And we should document the IANA source as well!

But we need to be aware that for many plugins devs and native implementers could take a while for this to implement.

And is going to impact the tests for the core modules? Don't think so, but just asking.

And once https://github.com/editorconfig/editorconfig-defaults has a stable schema it should be added there as well.

**Comment by ppalaga** [comment]
> we need to be aware that for many plugins devs and native implementers could take a while for this to implement.

That's normal and that's again a reason to introduce the `version` property https://github.com/editorconfig/editorconfig/issues/356 so that the processing tools can see whether the EditorConfig specification version the given `.editorconfig` file is targeting is not too new for them.

> And is going to impact the tests for the core modules? Don't think so, but just asking.

Yes, it is a good idea to add tests to check that the processing tools can handle the listed encodings. We currently have tests only for `charset = utf-8` and `charset = unset`. See https://github.com/editorconfig/editorconfig-core-test/search?q=charset&unscoped_q=charset

**Comment by kAlvaro** [comment]
I possibly misunderstood jedmao's comments (I'm just an end user, not involved or familiar with the EditorConfig project in any way) but I was under the impression that encoding support had deep technical implications in relates libraries. For instance, [ECLint](https://github.com/jedmao/eclint) project description says (emphasis mine):

> Validate or fix code that doesn't adhere to EditorConfig settings or **infer settings from existing code**.

If getting or agreeing on a normative list is all the difficulty, isn't IANA (plus existing non-IANA names for backwards compatibility) all we need? What'd the net benefit of supporting arbitrary names and spellings? Format expects `root = true` and you cannot use `root = on` or `root = yes` or `root: 1`. Format expects `end_of_line = lf` and you cannot use `Unix` or `U+000A`.

> .editorconfig processors should support **at least** the following common encoding names standardized by IANA

Not sure if this means that tools are allowed to support non-standard aliases (e.g. `CP1252` instead of `windows-1252`) or it's a fallback for esoteric encodings not even listed by IANA (like, I don't know, encodings supported by Russian mainfraimes in the 1950s). In either case, I'm inclined to keep it simple and clearly state the ["Preferred MIME Name" column](http://www.iana.org/assignments/character-sets/character-sets.xhtml) as only acceptable value.

**Comment by jedmao** [comment]
@kAlvaro I just want to be clear about some things, because my philosophies have evolved since creating ECLint.

1. I don't believe in the concept of ECLint any more. If I could go back, I wouldn't create it on the premise that it needs AST (knowledge of the language in question) to successfully "fix" some of the EditorConfig rules. As a result, the project has a `noop` for those rules. I didn't originally think this would be the case, but I was wrong.
1. In terms of encoding support, I believe the editor itself should provide such support. The purpose of any EditorConfig plugin would be to leverage that built-in functionality and switch it to the proper encoding. In the absence of such editor features, I would suggest extension developer to do nothing, as it would require some sophisticated logic and potentially dangerous implementations.

My ethos of recent is to minimize the "magic" that EditorConfig extensions might introduce and leverage built-in editor features wherever possible.

_Disclaimer: my views are mine alone and do not necessarily reflect the views of other EditorConfig members._

**Comment by ffes** [comment]
> The purpose of any EditorConfig plugin would be to leverage that built-in functionality and switch it to the proper encoding. In the absence of such editor features, I would suggest extension developer to do nothing, as it would require some sophisticated logic and potentially dangerous implementations.

Nicely said. This (slightly altered) should be project's motto and put on the site 😄

**Comment by ppalaga** [comment]
> isn't IANA (plus existing non-IANA names for backwards compatibility) all we need?

My proposal is IANA's MINE minus the ones unsupported by Java [1] plus the original EditorConfig ones plus UTF-16. Taking all IANA makes no sense, because no tool will be able to support them all. Taking these common encodings, we want to set the baseline for end users' expectations as well as a basic recommendation for EditorConfig implementors.

> Not sure if this means that tools are allowed to support non-standard aliases (e.g. `CP1252` instead of `windows-1252`) 

Yes, my idea was that the implementations are free use any aliases common in their ecosystem. For the Java based tools I work on, I'll simply take any encoding name accepted by Java's built in `Charset.forName(String)` method.

> or it's a fallback for esoteric encodings not even listed by IANA (like, I don't know, encodings supported by Russian mainfraimes in the 1950s). 

I see no reason why the implementors, if they see a need, should be prohibited by the EditorConfig spec to do so.

> In either case, I'm inclined to keep it simple and clearly state the ["Preferred MIME Name" column](http://www.iana.org/assignments/character-sets/character-sets.xhtml) as only acceptable value.

Preferred MIME Name does not contain any of the UTF encodings.

[1] https://docs.oracle.com/javase/8/docs/technotes/guides/intl/encoding.doc.html

**Comment by opensas** [comment]
I got a bit lost in the conversation, but from what I understand there's still no way to enforce  Windows-1252 from editorconfig, right???

**Comment by opensas** [comment]
I got a bit lost in the conversation, but from what I understand there's still no way to enforce  Windows-1252 from editorconfig, right???

**Comment by ppalaga** [comment]
> I would add the windows-* encodings to that list, they're not on under the "Preferred Mime Name" column either and very popular around here (central Europe).

I do not mind adding windows-874 and windows-1250 trough windows-1258 to the baseline list.

> Why allow implementations to interpret aliases as they see fit? Doesn't that go against the very core of having an .editorconfig file where everyone can agree on what one particular setting means?
> 
> I think the editorconfig project should have one blessed list of supported encoding names, where support is optional.

I like testable standards. Any implementation would be trivially compliant with a list of encoding names where every single item is optional. 

Requiring support for a short list of widely used encodings would be non trivial to test. Keeping the list short would increase the likehood for the implementations to pass. 

Anything beyond the baseline list should be best effort form the side of implementations so that users with special needs are not blocked by too narrow specs, if the tool(s) they happen to use support what they need.

**Comment by alshain** [comment]
> I do not mind adding windows-874 and windows-1250 trough windows-1258 to the baseline list.

Great! I'm also in favor of a baseline list. I didn't mean to say that support for every item on the "blessed list" should be optional.

I just read this:

> I see no reason why the implementors, if they see a need, should be prohibited by the EditorConfig spec to do so.

and for me, it reads as if this would give implementers leave to make whatever extensions to the EditorConfig spec that they wish, which would lead to many different and unofficial EditorConfig versions that are incompatible between each other. I don't think this would be beneficial for the community.

**Comment by concatime** [comment]
`charset = latin1` should refer to ISO-8859-1 (`latin1` is an [alias](//www.iana.org/assignments/character-sets/character-sets.xhtml#table-character-sets-1)). We also need something for **at least** `US-ASCII` and for `CP-1252`/`Windows-1252`. Also, avoid `ANSI` term, it’s [ambiguous](//sourceforge.net/p/notepad-plus/discussion/331753/thread/7e6624c4/).

**Comment by ams-tschoening** [comment]
Just to keep things alive, I stumbled across this problem as well, because the EditorConfig-plugin for Sublime Text doesn't support `windows-1252` as well. What's interesting is that if a file is already encoded in `windows-1252`, either the plugin or Sublime Text simply keeps that encoding. Don't know who guesses what when, but that's what I see and Sublime Text really shows the kept encoding in the status lines. A file not already containing `windows-1252`, but e.g. ASCII only, simply gets opened/stored using `UTF-8` if `windows-1252` is documented in `.editorconfig` instead of e.g. throwing an error.

And here's how the plugin author recognizes encodings:

	CHARSETS = {
		'latin1': 'Western (ISO 8859-1)',
		'utf-8': 'utf-8',
		'utf-8-bom': 'utf-8 with bom',
		'utf-16be': 'utf-16 be',
		'utf-16le': 'utf-16 le'
	}

https://github.com/sindresorhus/editorconfig-sublime/blob/master/EditorConfig.py#L24

		if charset in CHARSETS:
			view.set_encoding(CHARSETS[charset])

https://github.com/sindresorhus/editorconfig-sublime/blob/master/EditorConfig.py#L83

From my understanding, in case of an unsupported value, the plugin lets Sublime Text simply behave as it sees fit and which surely is using `UTF-8` in most cases.

So, in this example it seems to be trivial to add additional encodings to the whitelist and really not worth it discussing years about how much pressure this might apply to plugin authors. I would really be surprised if a large number of plugins wouldn't simply use some sort of easy mapping like above or forward all given encodings to their editor anyway, so that those can handle aliases and stuff.

So please add the [formerly mentioned encodings](https://github.com/editorconfig/editorconfig/issues/209#issuecomment-445241830) in addition to the main `windows-[...]` ones to the list, so that users can politely ask their plugin-authors to extend the support.

I was surprised that widespread encodings like `windows-1252` didn't have been supported right from the start, even though that is far more popular than `latin1`. From my experience, many people wrongly assume to work with `latin1`, when they really am using `windows-1252`.

**Comment by ksmutny** [comment]
I might be in a minority, but today I could have really use support for cp850 (yes, good old DOS). And was quite surprised only such a few charsets are supported until now.

Definitely +1 for extending support for multiple charsets.

**Comment by vstenvik** [comment]
> This brings up another question – how should we be determining that a document is of a particular encoding? Sure, the .editorconfig file says it **should** be a certain encoding, but that doesn't mean the document is already in that encoding.

This should be the concern of the editor and not the EditorConfig. Also, even if there are cases where the encoding of a file cannot be determined, setting what encoding the file should have, and what _new files_ should have still is valuable.

**Comment by ryandesign** [comment]
I am writing an AppleScript script. Every time the file is saved in Apple's Script Editor, it pretty-prints the script using tabs for indentation (including on otherwise blank lines) and converts the character encoding to MacRoman. (If I used non-MacRoman characters in my file, it would instead auto-convert it to UTF-16-LE, but I'm not doing that.) So that other editors might use the same conventions, I have expressed this in my .editorconfig with:

```ini
[*.applescript]
charset = macroman
indent_style = tab
trim_trailing_whitespace = false
```

This appears to be honored at least by BBEdit. I would appreciate `macroman` being an officially-sanctioned and -documented value for the editorconfig `charset` property.

See also #314.

**Comment by aka-nse** [comment]
In current [formal specification](https://spec.editorconfig.org/#supported-pairs), charset is defined such as:

> Set to latin1, utf-8, utf-8-bom, utf-16be or utf-16le to control the character set. Use of utf-8-bom is discouraged.

But [wiki page](https://github.com/editorconfig/editorconfig/wiki/Character-Set-Support) describe such as:

> We should decide which character sets to support for each/all plugins. Other character sets could be specified outside of this set and they would be ignored if not understood by the editor. This is simply a list of character sets that all plugins should attempt to support at a minimum.
>
> - UTF-8, BOM deprecated.
> - UTF-16, LE/BE, BOM unspecified.
> - Latin-1
> 
> If an implementation wishes to define a new character set, it is recommended that it consults outside authorities such as the WHATWG Encoding Standard and the IANA Character Set registry. Using a special prefix such as x- is neither required nor recommended, as unsupported values are simply ignored.

I feel there is difference of nuance between them.

Therefore I propose to update specification such as following:

> Following character sets are supported in every valid runtime:
>
> - "utf-8" (recommended)
> - "latin1"
> - "utf-8-bom" (discouraged)
> - "utf-16be"
> - "utf-16le"
> - "unset" (default, fallback: meaning to any charset is not specified)
>
> Furthermore runtimes can support more extensive charsets. In that case, the list of additional available charset is usually subset of [IANA Character Set registry](https://www.iana.org/assignments/character-sets/character-sets.xhtml). Unsupported options will be fallenback into "unset". To get information of  concrete options, please refer to the document of runtime to use.

**Comment by binki** [comment]
If the list of supported charsets is based on the editor/host environment, my concern is how to support different editors used by the same team with this behavior. If there is no way to specify a fallback list (similar to [CSS `font-family`](https://www.w3.org/TR/CSS2/fonts.html#font-family-prop)), there will be no way for a single `.editorconfig` file to support two editors which use different names for the same charset. One editor or the other will have to treat the value as `unset`.

To avoid this, editorconfig itself must either provide a fallback mechanism (so that the author of the `.editorconfig` file can support all of the editors required) or specify which standard to refer to. This way, it is actually possible to rely on `.editorconfig` even while using obscure encodings.

I think specifying an external standard such as IANA’s “preferred MIME Name” (with case-insensitive match) would be the best option. Then, it is clear that the responsibility for mapping names falls on plugin authors. Plugin authors for editors which are MIME-aware get free support. Other plugin authors will have to find a way to map to whatever the editor supports. But at least we won’t have a situation where one needs to use a different `.editorconfig` file depending on which editor they are using.

However, it feels like it is too late to fix this. No matter what approach is chosen (fallback support or referring to an external standard), all plugin authors will need to update their plugins.

**Comment by cxw42** [comment]
@ffes per your https://github.com/editorconfig/editorconfig/issues/519#issuecomment-2506204522 , should this be closed as duplicate of #467?

**Comment by tats-u** [comment]
@cxw42 It does not cover non-Latin encodings. If you close this this must change the title of that issue.

**Comment by florianb** [comment]
@ffes, @cxw42 yes please, let's move on to #467 - i think it's necessary to add more encodings.
